jax/_src/nn/functions.py

```diff
@@ -912,19 +912,25 @@ def dot_product_attention(
   Returns:
     An array of the attention output with the same shape as :code:`query`.
   """
+  original_shape = jnp.asarray(query).shape
+  def _preprocess_array(t):
+    if t is None:
+      return t
+    t = jnp.asarray(t)
+    return t[None, ...] if t.ndim == 3 else t
```

Comment:
Does it make sense to assert that `t.ndim` is 4 if it's not 3?

Reply:
Done.